flow-server/src/main/java/com/vaadin/flow/server/communication/rpc/AbstractRpcInvocationHandler.java

```diff
@@ -72,6 +77,110 @@ public Optional<Runnable> handle(UI ui, JsonObject invocationJson) {
+
+        if (!isPollEventInvocation(invocationJson)) {
+            return false;
+        }
+
+        if (!isPollingEnabledForUI(ui)) {
+            StringBuilder warning = new StringBuilder(
+                    "Ignoring Poll RPC for UI that does not have polling enabled.");
+            if (!VaadinService.getCurrent().getDeploymentConfiguration()
+                    .isProductionMode() && ui.getUIId() > 0) {
+                warning.append(" UI id: '{}'");
+            }
+            getLogger().warn(warning.toString(), ui.getUIId());
+            return false;
+        }
+
+        if (!isLegitimatePollEventInvocation(ui, invocationJson)) {
+            getLogger().warn("Ignoring Poll RPC for illegitimate invocation.");
+            return false;
+        }
+
+        return true;
+    }
+
+    private boolean isPollEventInvocation(JsonObject invocationJson) {
+        return invocationJson.hasKey(JsonConstants.RPC_EVENT_TYPE)
+                && PollEvent.DOM_EVENT_NAME.equalsIgnoreCase(
+                        invocationJson.getString(JsonConstants.RPC_EVENT_TYPE));
+    }
+
+    private boolean isPollingEnabledForUI(UI ui) {
+        return ui.getPollInterval() > 0
+                && ComponentUtil.hasEventListener(ui, PollEvent.class);
+    }
+
+    /**
+     * This method checks that a legitimate Poll Rpc invocation properties
+     * should contain only the following three <b>allowed</b> keys along with
+     * their values and nothing less or more:
+     * <ul>
+     * <li>{@link com.vaadin.flow.shared.JsonConstants#RPC_TYPE}</li>
+     * <li>{@link com.vaadin.flow.shared.JsonConstants#RPC_NODE}</li>
+     * <li>{@link com.vaadin.flow.shared.JsonConstants#RPC_EVENT_TYPE}</li>
+     * </ul>
+     * <p>
+     * As Rpc invocations of type polling would still be handled even while the
+     * UI is inert (due to server-modality) this will make sure that the request
+     * does not include any extra malicious payloads.
+     * <p>
+     * This method checks the existence of first two allowed keys as the
+     * {@link #isPollEventInvocation(JsonObject)} had already checked for the
+     * existence of the
+     * {@link com.vaadin.flow.shared.JsonConstants#RPC_EVENT_TYPE} before this
+     * method is called.
+     *
+     * @see #ignoreInertForRpcInvocation(UI, JsonObject)
+     *
+     * @param ui
+     *            the UI instance which the Rpc event is coming from.
+     * @param invocationJson
+     *            the Rpc invocation payload as Json.
+     * @return a boolean indicating whether the invocationJson is legitimate in
+     *         accordance with the UI instance.
+     */
+    private boolean isLegitimatePollEventInvocation(UI ui,
+            JsonObject invocationJson) {
+        int allowedKeysSize = 3;
+        String[] invocationKeys = invocationJson.keys();
+        if (invocationKeys == null
+                || invocationKeys.length != allowedKeysSize) {
+            return false;
+        }
```

Comment:
I would rather check that we have only the allowed amount and types of keys:
```suggestion
        List<String> allowedKeys = Arrays.asList(new String[]{JsonConstants.RPC_TYPE, JsonConstants.RPC_NODE, JsonConstants.RPC_EVENT_TYPE});
        List<String> invocationKeys = Arrays.asList(invocationJson.keys());
        if (!allowedKeys.containsAll(invocationKeys)) {
            return false;
        }
```

Reply:
Done.